packages/shared/src/components/PostOptionsReadingHistoryMenu.tsx

```diff
@@ -0,0 +1,168 @@
+
+const PortalMenu = dynamic(() => import('./fields/PortalMenu'), {
+  ssr: false,
+});
+
+export type PostOptionsReadingHistoryMenuProps = {
+  post: ReadHistoryPost;
+  onHiddenMenu?: () => unknown;
+  displayCopiedMessageFunc?: () => void;
+  onHideHistoryPost?: (postId: string) => unknown;
+  indexes: QueryIndexes;
+};
+
+/* eslint-disable no-param-reassign */
+const updateReadingHistoryPost =
+  (
+    queryClient: QueryClient,
+    historyQueryKey: QueryKey,
+    readHistoryPostUpdated: Partial<ReadHistoryPost>,
+    { page, edge }: QueryIndexes,
+  ): ((args: { id: string }) => Promise<() => void>) =>
+  async () => {
+    const oldReadingHistory =
+      queryClient.getQueryData<ReadHistoryInfiniteData>(historyQueryKey);
+    const newItems = [...oldReadingHistory.pages];
+    const history = newItems[page].readHistory.edges[edge].node.post;
+    newItems[page].readHistory.edges[edge].node.post = {
+      ...history,
+      ...readHistoryPostUpdated,
+    };
+    queryClient.setQueryData<ReadHistoryInfiniteData>(historyQueryKey, {
+      ...oldReadingHistory,
+      pages: newItems,
+    });
+    return () => {
+      queryClient.setQueryData<ReadHistoryInfiniteData>(
+        historyQueryKey,
+        oldReadingHistory,
+      );
+    };
+  };
+
+const getBookmarkIconAndMenuText = (bookmarked: boolean) => {
+  if (bookmarked) {
+    return (
+      <>
+        <MenuIcon Icon={BookmarkIconFilled} />
+        Remove from bookmarks
+      </>
+    );
+  }
+  return (
+    <>
+      <MenuIcon Icon={BookmarkIcon} />
+      Save to bookmarks
+    </>
+  );
+};
+
+export default function PostOptionsReadingHistoryMenu({
+  post,
+  onHiddenMenu,
+  onHideHistoryPost,
+  indexes,
+}: PostOptionsReadingHistoryMenuProps): ReactElement {
+  const { user } = useContext(AuthContext);
+  const { trackEvent } = useContext(AnalyticsContext);
+  const queryClient = useQueryClient();
+  const historyQueryKey = ['readHistory', user?.id];
+
+  const [, copyLink] = useCopyLink(() => post.commentsPermalink);
+
+  const trackEventFunc = (eventName: string) => {
+    return () =>
+      trackEvent(
+        postAnalyticsEvent(eventName, post, {
+          extra: { origin: 'reading history post menu' },
+        }),
+      );
+  };
+
+  const onShareOrCopyLink = useShareOrCopyLink({
+    link: post?.commentsPermalink,
+    text: post?.title,
+    copyLink,
+    trackEvent: trackEventFunc('share post'),
+  });
+
+  const { bookmark, removeBookmark } = useBookmarkPost({
+    onBookmarkMutate: updateReadingHistoryPost(
+      queryClient,
+      historyQueryKey,
+      {
+        bookmarked: true,
+      },
+      indexes,
+    ),
+    onRemoveBookmarkMutate: updateReadingHistoryPost(
+      queryClient,
+      historyQueryKey,
+      {
+        bookmarked: false,
+      },
+      indexes,
+    ),
+  });
+
+  const onBookmarkReadingHistoryPost = async (): Promise<void> => {
+    const bookmarked = post?.bookmarked;
+    const trackEventFunction = trackEventFunc(
```

Comment:
Nice one of reusing this!
I'm personally not a fan of `func` but it's not breaking so leaving it up to @idoshamun  if he wants that changed

Reply:
I am not a fan either! 😅

Reply:
Instead of a function that returns a function, we can do a regular function that calls `trackEvent`, and then for `useShareOrCopyLink` we can use the `bind` function. Something like `trackEventFunc.bind(null, 'share post')` will generate a "new" function that receives no arguments. What do you think?

See here more info about it:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_objects/Function/bind

Reply:
@idoshamun Out of interest what's the main argument to do that compared to how we had it?

Reply:
@rebelchris I tried to provide a better solution following your comment that you are not a fan of the `func` thingy. So I provided an alternative that kind of suits this case better so we won't have to generate a function and then call it in this case.

Reply:
Oh right i only meant the shortcut name "func" i hate seeing those for some reason 😂🤯

Reply:
Oh ok then. Anyway, whatever works for you two. I don't mind just finding the `bind` solution a bit cleaner.